tests/phpunit/tests/functions/canonicalCharset.php

```diff
@@ -10,45 +10,47 @@
  * @covers ::_canonical_charset
  */
 class Tests_Functions_CanonicalCharset extends WP_UnitTestCase {
-
-	public function test_utf_8_lower() {
-		$this->assertSame( 'UTF-8', _canonical_charset( 'utf-8' ) );
-	}
-
-	public function test_utf_8_upper() {
-		$this->assertSame( 'UTF-8', _canonical_charset( 'UTF-8' ) );
-	}
-
-	public function test_utf_8_mixxed() {
-		$this->assertSame( 'UTF-8', _canonical_charset( 'Utf-8' ) );
-	}
-
-	public function test_utf_8() {
-		$this->assertSame( 'UTF-8', _canonical_charset( 'UTF8' ) );
-	}
-
-	public function test_iso_lower() {
-		$this->assertSame( 'ISO-8859-1', _canonical_charset( 'iso-8859-1' ) );
-	}
-
-	public function test_iso_upper() {
-		$this->assertSame( 'ISO-8859-1', _canonical_charset( 'ISO-8859-1' ) );
-	}
-
-	public function test_iso_mixxed() {
-		$this->assertSame( 'ISO-8859-1', _canonical_charset( 'Iso8859-1' ) );
-	}
-
-	public function test_iso() {
-		$this->assertSame( 'ISO-8859-1', _canonical_charset( 'ISO8859-1' ) );
-	}
-
-	public function test_random() {
-		$this->assertSame( 'random', _canonical_charset( 'random' ) );
+	/**
+	 * Ensures that charset variants for common encodings normalize to the expected form.
+	 *
+	 * @ticket 61182
+	 *
+	 * @dataProvider data_charset_normalizations
+	 *
+	 * @param string $given_charset      Potential charset provided by user.
+	 * @param string $normalized_charset Expected normalized form of charset.
+	 */
+	public function test_properly_normalizes_charset_variants( $given_charset, $normalized_charset ) {
+		$this->assertSame(
+			$normalized_charset,
+			_canonical_charset( $given_charset ),
+			'Did not properly transform the provided charset into its normalized form.'
+		);
 	}
 
-	public function test_empty() {
-		$this->assertSame( '', _canonical_charset( '' ) );
+	/**
+	 * Data provider.
+	 *
+	 * @return array[].
+	 */
+	public static function data_charset_normalizations() {
+		return array(
+			// UTF-8 family.
+			array( 'UTF-8', 'UTF-8' ),
+			array( 'Utf-8', 'UTF-8' ),
+			array( 'Utf-8', 'UTF-8' ),
+			array( 'UTF8', 'UTF-8' ),
+
+			// ISO-8859-1 family.
+			array( 'iso-8859-1', 'ISO-8859-1' ),
+			array( 'ISO-8859-1', 'ISO-8859-1' ),
+			array( 'Iso-8859-1', 'ISO-8859-1' ),
+			array( 'ISO8859-1', 'ISO-8859-1' ),
+
+			// Other charset slugs should not be adjusted.
```

Comment:
I confirmed this is working correctly, but some substring (we do have `''` test) and superstring tests might be good:

```suggestion
			// Other charset slugs should not be adjusted.
			array( 'ISO8859-128', 'ISO8859-128' ),
			array( 'ISO', 'ISO' ),
			array( 'UTF8080', 'UTF8080' ),
			array( 'UTF', 'UTF' ),
```

Note that those _would be_ a bug with the `substr_compare` implementation which returns the index of the string. `strcasecmp` returns non-0 to indicate the strings don't match.

Reply:
> those would be a bug with the substr_compare implementation which returns the index of the string

not sure I follow. `substr_compare()` doesn't return the index of a string. it starts at an offset in the haystack, and compares the given number of characters. if the string lengths are equal then this assures that the strings are identical, up to case sensitivity.

Reply:
It's not that well behaved, especially before php 8.2. It's… very strange. But we're fine since we're not using `substr_compare`:

![Screenshot 2024-05-15 at 14 54 22](https://github.com/WordPress/wordpress-develop/assets/841763/c4f33712-fc6a-4c51-a1e2-ddbae5928594)

Reply:
I think we're talking past each other, as I'm not sure what the code snippet is supposed to suggest.

We don't care about the value of the response, only if it's less than, greater than, or equal to `0`. For our purposes, we only care about `0` and equality, which it does fine.

Missing from all your examples is the critical `$length` parameter. The length parameter establishes that it will stop comparing after that many bytes, or whenever one of the string ends, if one ends before the `$length` has been reached.

<table><tr><td>
<img width="595" alt="Screenshot 2024-05-15 at 10 19 39 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/39314c42-6a54-4d8f-a19d-36230c1df326"><td>
<img width="464" alt="Screenshot 2024-05-15 at 10 19 53 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/e9fbd419-ae74-48d5-a43b-5bb417895860"></tr></table>

Does this make more sense?

Reply:
Before, the implementation looked like this where `$charset` with extra characters would match:

```php
$charset = 'iso-8859-1337';
var_dump( 0 === substr_compare( 'iso-8859-1', $charset, 0, 10, true ) ); // bool(true)
```

That's what had me thinking along this line, but again, it's not a problem because the final implementation is using functions that don't have that problem.

We can drop it 🙂

Reply:
> We can drop it 🙂

yeah I realize it's not important, but I did originally have a length check, and a note about that. more or less at this point I want to ensure I'm not overlooking something, and that we both reach a shared understanding, if we want.

> if the string lengths are equal then this assures that the strings are identical, up to case sensitivity.

so there is no bug because if it compares 10 characters and the charset _is_ only 10 characters I don't think it's possible to be wrong.

https://github.com/WordPress/wordpress-develop/pull/6535/commits/a7d7696e594742d6144f7189dd920416a996bd45#diff-5526e2d644a04f0c466d6c749542fdb7952021784e7b1583ff45f9d12601ef52R7497-R7498

```php
if (
	10 === $charset_length &&
	substr_compare( $charset, 'iso-8859-1', 0, 10, true )
)
```

Reply:
You are correct 😆